### DISCUSSION TITLE
multi-species species counts

### DISCUSSION THREAD

**Discussion by svshepherd** [pyinat/pyinaturalist#550] [Q&A] [answered]
Howdy, everyone. I'm working on a function to get a list of species to look for in a given locale at the current moment.  I'd like to change the sort order based on how unique a species is: normalizing the number of observations here-and-now by the total similar observations in a place or in a time.  However, when many species are returned, this is slow and seems to drag on inat's servers, sometimes throwing errors.  

I've already gathered the results for here-and-now as a pandas dataframe.  Rather than moving taxon-by-taxon with iterrows(), I tried something like the following, which returned no results. Is there a better approach?  

`inat.get_observation_species_counts(taxon_id=[results['taxon.id'].to_list()], term_id=12, term_value_id=13, lat=37, lng=-77, radius=25)`

**Comment by JWCook** [comment]
Since I don't have your dataframe contents to test with, can you show me the debug log output for this search? You can enable it with `pyinaturalist.enable_logging()`. You can also do that with `logging.basicConfig()`; this just adds some useful defaults for nicer log formatting.
```py
from pyinaturalist import enable_logging
enable_logging('DEBUG')
```

That will show me the full HTTP request being sent.

**Reply by svshepherd** [reply]
![image](https://github.com/pyinat/pyinaturalist/assets/47264767/92339f8b-b19b-423a-a49f-119c05f6f864)

**Reply by JWCook** [reply]
Oh! You're passing a nested list for taxon IDs. Change that argument to:
```py
taxon_id=results['taxon.id'].to_list()
```

**Reply by svshepherd** [reply]
Ah, yes, foolish mistake on my part -- but it looks similar after correcting.

![image](https://github.com/pyinat/pyinaturalist/assets/47264767/3013525f-d29e-434d-8a66-4827d462d22c)

**Reply by JWCook** [reply]
Is it possible that there just aren't any results matching those filters? Without filtering by `taxon_id`, there are only 66 results in that location with that annotation value. It's also possible that the list of taxon IDs is longer than what the iNat API accepts, although usually that would result in an error response.

If you can paste or attach the full list of taxon IDs you have, I can help debug further.

**Reply by svshepherd** [reply]
Thanks much!  Let me poke at this for a bit and see if I can come up with some answers on my own. I appreciate your help!

**Comment by svshepherd** [comment, marked as answer]
Not quite sure what the magic change was -- I suspect I was over some limit in terms of species requested -- but I think I've got things working pretty well now.  Thank you! The multi-species request is much more performant than going one at a time.  My little project currently has two functions, one for finding and formatting names, and one for showing species to keep an eye out for in the current place/time.  I'm open to feedback and you're welcome to use anything you like.

https://github.com/svshepherd/pynat
